### PR TITLE
refactor: add scopes to credentials for provisioners

### PR DESCRIPTION
**Pull request by man8pr**
## What this PR changes/adds

IamService provides methods for getting credentials from token or service account, using optionally scopes.

## Why it does that

Constraining the scope of the credentials improves the security.

### CODE REVIEW

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java [hunk 1]
```diff
@@ -97,6 +99,45 @@ public GcpAccessToken createAccessToken(GcpServiceAccount serviceAccount) {
         }
     }
 
+    @Override
+    public GoogleCredentials getCredentials(GcpAccessToken accessToken) {
```
Comment (ndr-brt) [inline extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java:103]: please test this method

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java [hunk 2]
```diff
@@ -97,6 +99,45 @@ public GcpAccessToken createAccessToken(GcpServiceAccount serviceAccount) {
...
+        return GoogleCredentials.create(
+                new AccessToken(accessToken.getToken(), new Date(accessToken.getExpiration()))
+        );
+    }
+
+    @Override
+    public GoogleCredentials getCredentials(String serviceAccountName, String... scopes) {
```
Comment (ndr-brt) [inline extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java:110]: please test this method

**Review by ndr-brt — CHANGES_REQUESTED** [review]
please add coverage on the new code and avoid duplication

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (man8pr) [reply extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java:110]: Correct, the plan is to use IamService.getCredentials in other classes, I will make the change visible in this PR

[the remark this answers is not in the document]
Comment (man8pr) [reply extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java]: Done

File: extensions/data-plane/data-plane-google-storage/src/main/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSinkFactory.java [hunk 3]
```diff
@@ -93,16 +90,9 @@ private Storage createStorageClient(String keyName) {
...
-                            new Date(gcsAccessToken.getExpiration()))
-            );
+            googleCredentials = iamService.getCredentials(gcsAccessToken);
         } else {
-            try {
-                googleCredentials = GoogleCredentials.getApplicationDefault();
-            } catch (IOException e) {
-                throw new GcpException("Error while getting the default credentials.", e);
-            }
+            googleCredentials = iamService.getCredentials(null, "https://www.googleapis.com/auth/devstorage.read_write");
```
[the remark this answers is not in the document]
Comment (man8pr) [reply extensions/data-plane/data-plane-google-storage/src/main/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSinkFactory.java]: Done

[the remark this answers is not in the document]
Comment (man8pr) [reply extensions/data-plane/data-plane-google-storage/src/main/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSinkFactory.java]: Done

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java (hunk 1 above)
Comment (ndr-brt) [reply extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java:103]: At least one test that passes a `GcpAccessToken` and verifies that in output we have a `GoogleCredential` containing the token composed in a certain way, a classic unit test

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java [hunk 4]
```diff
@@ -174,9 +215,12 @@ private Supplier<IamCredentialsClient> defaultIamCredentialsClientSupplier() {
 
     private record ApplicationDefaultCredentials(Monitor monitor) implements AccessTokenProvider {
         @Override
-        public GcpAccessToken getAccessToken() {
+        public GcpAccessToken getAccessToken(String... scopes) {
             try {
-                var credentials = GoogleCredentials.getApplicationDefault().createScoped(IamScopes.CLOUD_PLATFORM);
+                var credentials = GoogleCredentials.getApplicationDefault();
+                if (scopes != null) {
```
[the remark this answers is not in the document]
Comment (man8pr) [reply extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java]: Updated the ApplicationDefaultCredentials interface to CredentialsUtil, with more methods, that are basically 1-line invocations to GCP APIs

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java (hunk 1 above)
Comment (man8pr) [reply extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java:107]: Thanks for the offline clarification, I have used mocks to test the method

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java (hunk 2 above)
Comment (man8pr) [reply extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java]: Done

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/service/BigQueryFactoryImpl.java [hunk 5]
```diff
@@ -25,36 +23,22 @@
 import org.eclipse.edc.spi.monitor.Monitor;
 
 import java.io.IOException;
-import java.util.Arrays;
 
 public class BigQueryFactoryImpl implements BigQueryFactory {
     private final GcpConfiguration gcpConfiguration;
     private final Monitor monitor;
```
Comment (ndr-brt) [inline extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/service/BigQueryFactoryImpl.java:29]: monitor is not used anymore

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/service/BigQueryFactoryImpl.java [hunk 6]
```diff
@@ -25,36 +23,22 @@
...
 
-    public BigQueryFactoryImpl(GcpConfiguration gcpConfiguration, Monitor monitor) {
+    public BigQueryFactoryImpl(GcpConfiguration gcpConfiguration, Monitor monitor, IamService iamService) {
         this.gcpConfiguration = gcpConfiguration;
         this.monitor = monitor;
+        this.iamService = iamService;
     }
 
     @Override
     public BigQuery createBigQuery(GcpServiceAccount serviceAccount) throws IOException {
```
Comment (ndr-brt) [inline extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/service/BigQueryFactoryImpl.java:39]: no `IOException` is thrown by this method

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/CredentialsUtil.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+package org.eclipse.edc.gcp.iam;
+
+import com.google.auth.oauth2.GoogleCredentials;
+import org.eclipse.edc.gcp.common.GcpServiceAccount;
+
+/**
+ * Interface for credentials providing access tokens.
+ */
+public interface CredentialsUtil {
```
Comment (ndr-brt) [inline extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/CredentialsUtil.java:23]: please avoid the `Util` word that could lead to undefined scope and "trash bin classes", `CredentialsManager` could be a more appropriate name

File: extensions/data-plane/data-plane-google-storage/src/test/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSinkFactoryTest.java [hunk 8]
```diff
@@ -32,51 +38,64 @@
...
+    void instantiateMocksAndVault() {
+        vault = mock();
+        iamService = mock();
+        factory = new GcsDataSinkFactory(
+                mock(ExecutorService.class),
+                mock(Monitor.class),
+                vault,
+                typeManager,
+                iamService
+        );
```
Comment (ndr-brt) [inline extensions/data-plane/data-plane-google-storage/src/test/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSinkFactoryTest.java:62]: nit: these assignments can be moved to the field declarations

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java [hunk 9]
```diff
@@ -172,18 +209,40 @@ private Supplier<IamCredentialsClient> defaultIamCredentialsClientSupplier() {
         }
     }
 
-    private record ApplicationDefaultCredentials(Monitor monitor) implements AccessTokenProvider {
+    private record DefaultCredentialsUtil(Monitor monitor) implements CredentialsUtil {
```
Comment (ndr-brt) [inline extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java:212]: please extract to a separate class and document

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java [hunk 10]
```diff
@@ -172,18 +209,40 @@ private Supplier<IamCredentialsClient> defaultIamCredentialsClientSupplier() {
...
-                return null;
+                monitor.severe("Cannot get refresh the credentials", ioException);
+                throw new GcpException(ioException);
             }
         }
+
+        @Override
+        public GoogleCredentials createScoped(GoogleCredentials credentials, String... scopes) {
+            return credentials.createScoped(scopes);
+        }
```
Comment (ndr-brt) [inline extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java:236]: this method doesn't look like it needs to stay here, as the behavior for create scoped credentials is already provided by the `GoogleCredentials` class and it's pretty straightforward

**Review by ndr-brt — CHANGES_REQUESTED** [review]
there's an unneded `allDependencies` file that needs to be removed from this PR

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java (hunk 10 above)
Comment (man8pr) [reply extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java]: Done

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java (hunk 9 above)
Comment (man8pr) [reply extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/IamServiceImpl.java]: Done

File: extensions/data-plane/data-plane-google-storage/src/test/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSinkFactoryTest.java (hunk 8 above)
Comment (man8pr) [reply extensions/data-plane/data-plane-google-storage/src/test/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSinkFactoryTest.java]: Done

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/CredentialsUtil.java (hunk 7 above)
Comment (man8pr) [reply extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/iam/CredentialsUtil.java]: Done

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/service/BigQueryFactoryImpl.java (hunk 6 above)
Comment (man8pr) [reply extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/service/BigQueryFactoryImpl.java]: Done

File: extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/service/BigQueryFactoryImpl.java (hunk 5 above)
Comment (man8pr) [reply extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/service/BigQueryFactoryImpl.java]: Done

**Comment by man8pr** [comment]
> there's an unneded `allDependencies` file that needs to be removed from this PR

Removed, thanks for the finding!

**Review by ndr-brt — APPROVED** [review]

**Pull request merged**